[AppKit Gestures] Refactor wheel input handling to support NSPanGestureRecognizer

aprotyas opened the pull request:
#### 78ad297be7f772ab584bb4a1110e2cea163c57af
<pre>
[AppKit Gestures] Refactor wheel input handling to support NSPanGestureRecognizer
<a href="https://bugs.webkit.org/show_bug.cgi?id=302494">https://bugs.webkit.org/show_bug.cgi?id=302494</a>
<a href="https://rdar.apple.com/163242714">rdar://163242714</a>

Reviewed by Lily Spiniolas and Wenson Hsieh.

This patch adds infrastructure to support scrolling web views using
NSPanGestureRecognizer, modernizing our gesture handling to adopt
AppKit&apos;s gesture recognizer APIs in place of lower-level event handling.

The implementation introduces WKPanGestureController, which translates
NSPanGestureRecognizer events into WebWheelEvents that can drive the
existing scrolling machinery. This behavior change is gated behind a new
UseAppKitGestures internal preference.

Test coverage will be added in a follow up patch.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/Scrollbar.h:
(WebCore::Scrollbar::pixelsPerLineStep): Drive-by: make this constexpr.
* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Source/WebKit/Shared/NativeWebWheelEvent.h:
* Source/WebKit/Shared/mac/NativeWebWheelEventMac.mm:
(WebKit::NativeWebWheelEvent::NativeWebWheelEvent):
  Added constructor to create NativeWebWheelEvent from WebWheelEvent,
  enabling gesture-synthesized wheel events without a backing NSEvent.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/mac/WKPanGestureController.h: Added.
* Source/WebKit/UIProcess/mac/WKPanGestureController.mm: Added.
(translationInView):
(toWheelEventPhase):
(-[WKPanGestureController configureForScrolling:]):
(-[WKPanGestureController initWithPage:viewImpl:]):
(-[WKPanGestureController panGestureRecognized:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::m_flagsChangedEventMonitorTrackingArea):
(WebKit::WebViewImpl::configurePanGestureRecognizerIfNeeded):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/303084@main">https://commits.webkit.org/303084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71dda3dc9e66285cd5adb137a2d6219bea805fc9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131180 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3510 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42144 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138622 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82886 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/747fea51-60b1-4a12-b66e-1fb31ab50bf0) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3503 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3351 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99943 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67702 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6af1eccf-657d-4c1f-b779-4625eaa5255a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134126 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2499 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117428 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80642 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/69d5f682-0f4f-4c14-91d3-faaa91434399) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2419 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/117 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81871 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123202 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111034 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35545 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141120 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129634 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3255 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36073 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108464 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3300 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2908 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108406 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2440 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113760 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56314 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20415 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3317 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32184 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162651 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3139 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66724 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40612 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3339 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3247 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -21871,9 +21877,7 @@
-			platformFilters = (
-				ios,
-			);
+			platformFilter = ios;
```
I assume Xcode automatically made this change? It's probably fine to keep, if so.

aprotyas replied:
Right

whsieh commented on Source/WebKit/UIProcess/mac/WKPanGestureController.h:
```diff
@@ -0,0 +1,49 @@
+@interface WKPanGestureController : NSObject <NSGestureRecognizerDelegate>
+
+- (instancetype)initWithPage:(std::reference_wrapper<WebKit::WebPageProxy>)page view:(WKWebView *)view viewImpl:(std::reference_wrapper<WebKit::WebViewImpl>)viewImpl;
+- (void)setPanGestureRecognizer:(NSPanGestureRecognizer *)gesture;
```
Nit - maybe instead of a setter, you could either:
- Pass in the pan gesture recognizer upon initialization (since it doesn't seem to change afterwards), or
- Have the `WKPanGestureController` internally create and manage the pan gesture (this would also allow you to avoid exposing `-panGestureRecognized:` in the header)

aprotyas replied:
Good idea, yes it doesn't change afterwards.

I will have `WKPanGestureController` create the gesture, and maybe expose it to `WebViewImpl` if needed.

aprotyas replied:
> I will have `WKPanGestureController` create the gesture, and maybe expose it to `WebViewImpl` if needed.

I went with this approach here https://github.com/WebKit/WebKit/commit/1ba0c87bc6cf35ebc173b6df4d34a0a6c8b783ba, have not exposed it yet because there hasn't been a need.

pxlcoder commented on Source/WebKit/UIProcess/mac/WKPanGestureController.mm:
```diff
@@ -0,0 +1,178 @@
+#import <wtf/UUID.h>
+
+using namespace WebCore;
+using namespace WebKit;
```
I thought we generally preferred to avoid `using namespace` in the global namespace. It can cause issues with unified sources.

https://commits.webkit.org/194602@main

aprotyas replied:
Yes that's a good point, I will undo this.

aprotyas replied:
Stopped using `using namespace` in https://github.com/WebKit/WebKit/commit/1ba0c87bc6cf35ebc173b6df4d34a0a6c8b783ba.

pxlcoder commented on Source/WebKit/UIProcess/mac/WKPanGestureController.mm:
```diff
@@ -0,0 +1,178 @@
+    RetainPtr<NSPanGestureRecognizer> _panGestureRecognizer;
+}
+
+- (instancetype)initWithPage:(std::reference_wrapper<WebPageProxy>)page view:(WKWebView *)view viewImpl:(std::reference_wrapper<WebViewImpl>)viewImpl
```
I think you can omit the `view` parameter. `view()` is a public method on `WebViewImpl`.

aprotyas replied:
Makes sense

aprotyas replied:
Removed in https://github.com/WebKit/WebKit/commit/1ba0c87bc6cf35ebc173b6df4d34a0a6c8b783ba

aprotyas committed:
[AppKit Gestures] Refactor wheel input handling to support NSPanGestureRecognizer
https://bugs.webkit.org/show_bug.cgi?id=302494
rdar://163242714

Reviewed by Lily Spiniolas and Wenson Hsieh.

This patch adds infrastructure to support scrolling web views using
NSPanGestureRecognizer, modernizing our gesture handling to adopt
AppKit's gesture recognizer APIs in place of lower-level event handling.

The implementation introduces WKPanGestureController, which translates
NSPanGestureRecognizer events into WebWheelEvents that can drive the
existing scrolling machinery. This behavior change is gated behind a new
UseAppKitGestures internal preference.

Test coverage will be added in a follow up patch.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/Scrollbar.h:
(WebCore::Scrollbar::pixelsPerLineStep): Drive-by: make this constexpr.
* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Source/WebKit/Shared/NativeWebWheelEvent.h:
* Source/WebKit/Shared/mac/NativeWebWheelEventMac.mm:
(WebKit::NativeWebWheelEvent::NativeWebWheelEvent):
  Added constructor to create NativeWebWheelEvent from WebWheelEvent,
  enabling gesture-synthesized wheel events without a backing NSEvent.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/mac/WKPanGestureController.h: Added.
* Source/WebKit/UIProcess/mac/WKPanGestureController.mm: Added.
(translationInView):
(toWheelEventPhase):
(-[WKPanGestureController configureForScrolling:]):
(-[WKPanGestureController initWithPage:viewImpl:]):
(-[WKPanGestureController panGestureRecognized:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::m_flagsChangedEventMonitorTrackingArea):
(WebKit::WebViewImpl::configurePanGestureRecognizerIfNeeded):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/303084@main
